libs/core/langchain_core/runnables/graph.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Dict, List, NamedTuple, Optional, Type, Union
+from uuid import uuid4
+
+from langchain_core.pydantic_v1 import BaseModel
+from langchain_core.runnables.base import Runnable
+from langchain_core.runnables.graph_draw import draw
+
+
+class Edge(NamedTuple):
+    source: str
+    target: str
+
+
+class Node(NamedTuple):
+    id: str
+    data: Union[Type[BaseModel], Runnable]
+
+
+@dataclass
+class Graph:
+    nodes: Dict[str, Node] = field(default_factory=dict)
+    edges: List[Edge] = field(default_factory=list)
+
+    def next_id(self) -> str:
+        return uuid4().hex
+
+    def add_node(self, data: Union[Type[BaseModel], Runnable]) -> Node:
+        """Add a node to the graph and return it."""
+        node = Node(id=self.next_id(), data=data)
+        self.nodes[node.id] = node
+        return node
+
+    def remove_node(self, node: Node) -> None:
+        """Remove a node from the graphm and all edges connected to it."""
+        self.nodes.pop(node.id)
+        self.edges = [
+            edge
+            for edge in self.edges
+            if edge.source != node.id and edge.target != node.id
+        ]
+
+    def add_edge(self, source: Node, target: Node) -> Edge:
+        """Add an edge to the graph and return it."""
+        if source.id not in self.nodes:
+            raise ValueError(f"Source node {source.id} not in graph")
+        if target.id not in self.nodes:
+            raise ValueError(f"Target node {target.id} not in graph")
+        edge = Edge(source=source.id, target=target.id)
+        self.edges.append(edge)
+        return edge
+
+    def extend(self, graph: Graph) -> None:
+        """Add all nodes and edges from another graph.
+        Note this doesn't check for duplicates, nor does it connect the graphs."""
+        self.nodes.update(graph.nodes)
+        self.edges.extend(graph.edges)
+
+    def first_node(self) -> Optional[Node]:
```

Comment:
nit: source_node and sink_node are more intuitive to me than first_node and last_node but maybe just me